lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,872 @@
+//===-- lib/lower/ast-builder.cc ------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/lower/ASTBuilder.h"
+#include "flang/parser/parse-tree-visitor.h"
+#include <cassert>
+#include <utility>
+
+/// Build a light-weight AST to help with lowering to FIR.  The AST will
+/// capture pointers back into the parse tree, so the parse tree data structure
+/// may <em>not</em> be changed between the construction of the AST and all of
+/// its uses.
+///
+/// The AST captures a structured view of the program.  The program is a list of
+/// units.  Function like units will contain lists of evaluations.  Evaluations
+/// are either statements or constructs, where a construct contains a list of
+/// evaluations.  The resulting AST structure can then be used to create FIR.
+
+namespace Br = Fortran::lower;
+namespace Co = Fortran::common;
+namespace L = llvm;
+namespace Pa = Fortran::parser;
+
+using namespace Fortran;
+using namespace Br;
+
+namespace {
+
+/// The instantiation of a parse tree visitor (Pre and Post) is extremely
+/// expensive in terms of compile and link time, so one goal here is to limit
+/// the bridge to one such instantiation.
+class ASTBuilder {
+public:
+  ASTBuilder() {
+    pgm = new AST::Program;
```

Comment:
Where is this freed? Why not use a smart pointer or store values in parents rather than pointers?

Reply:
This was actually freed in the bridge code that will come in the "Add bridge from AST to FIR" was mentioned here: http://lists.llvm.org/pipermail/flang-dev/2020-January/000162.html

However, a smart pointer sounds a good idea here thanks.
`ASTBuilder::pgm` is now an `std::unique_ptr` and `createAST` returns it.